src/content.rs

```diff
@@ -515,3 +835,327 @@ impl LineCapStyle {
+    /// be used for function shadings. The matrix has six fields. Will otherwise
+    /// default to the identity matrix.
+    pub fn matrix(&mut self, matrix: impl IntoIterator<Item = f32>) -> &mut Self {
+        self.dict.key(Name(b"Matrix")).array().typed().items(matrix);
+        self
+    }
+
+    /// Write the `/Function` attribute.
+    ///
+    /// Sets the 2-in function to use for shading. Required.
+    pub fn function(&mut self, function: Ref) -> &mut Self {
+        self.dict.pair(Name(b"Function"), function);
+        self
+    }
+
+    /// Write the `/Coords` attribute.
+    ///
+    /// Sets the coordinates of the start and end of the axis in terms of the
+    /// target coordinate system. Required for axial (4 items) and radial (6
+    /// items; centers and radii) shadings.
+    pub fn coords(&mut self, coords: impl IntoIterator<Item = f32>) -> &mut Self {
+        self.dict.key(Name(b"Coords")).array().typed().items(coords);
+        self
+    }
+
+    /// Write the `/Extend` attribute.
+    ///
+    /// Set whether the shading should extend beyond either side of the axis /
+    /// circles. Can be used for axial and radial shadings.
+    pub fn extend(&mut self, extend: (bool, bool)) -> &mut Self {
+        self.dict
+            .key(Name(b"Extend"))
+            .array()
+            .typed()
+            .items([extend.0, extend.1]);
+        self
+    }
+}
+
+deref!('a, Shading<'a> => Dict<'a, IndirectGuard>, dict);
+
+/// What kind of shading to use.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+#[non_exhaustive]
+pub enum ShadingType {
+    /// The function specifies the color for each point in the domain.
+    Function,
+    /// The function specifies the color for each point on a line.
+    Axial,
+    /// The function specifies the color for each circle between two nested circles.
+    Radial,
+}
+
+impl ShadingType {
+    fn to_int(self) -> i32 {
+        match self {
+            Self::Function => 1,
+            Self::Axial => 2,
+            Self::Radial => 3,
+        }
+    }
+}
+
+/// Writer for a _tiling pattern stream_.
+pub struct ShadingPattern<'a> {
+    dict: Dict<'a, IndirectGuard>,
+}
+
+impl<'a> ShadingPattern<'a> {
+    pub(crate) fn start(obj: Obj<'a, IndirectGuard>) -> Self {
+        let mut dict = obj.dict();
+        dict.pair(Name(b"Type"), Name(b"Pattern"));
+        dict.pair(Name(b"PatternType"), PatternType::Shading.to_int());
+        Self { dict }
+    }
+
+    /// Write the `/Shading` attribute.
+    ///
+    /// Sets the shading object to use. Required.
+    pub fn shading(&mut self, shading: Ref) -> &mut Self {
+        self.dict.pair(Name(b"Shading"), shading);
+        self
+    }
+
+    /// Write the `/Matrix` attribute.
+    ///
+    /// Sets the matrix (6 items) to use for the pattern. Defaults to the
+    /// identity matrix.
+    pub fn matrix(&mut self, matrix: impl IntoIterator<Item = f32>) -> &mut Self {
```

Comment:
At this point, I guess, all matrices should be `[f32; 6]`. Probably the text matrix function should also be updated to take an array.